junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,544 @@
+              val cond = gen.mkMethodCall(idParamSym, nme.equals_, Nil, List(headMethodId))
+              val instance = gen.mkAttributedIdent(instanceParamSym)
+              val castedInstance = gen.mkAttributedCast(instance, refClassSym.tpe)
+              val thenp = gen.mkMethodCall(castedInstance, methodSymbol, Nil, Nil)
+              If(cond, thenp, acc)
+          }
+        }
+
+        val invokeJUnitMethodRhs =
+          atOwner(invokeJUnitMethodSym)(typer.typed(invokeJUnitMethodRhs0))
+        val istanceParam = newValDef_compat(instanceParamSym, EmptyTree)()
+        val idParam = newValDef_compat(idParamSym, EmptyTree)()
+        typer.typedDefDef(newDefDef_compat(invokeJUnitMethodSym, invokeJUnitMethodRhs)(
+          vparamss = List(List(istanceParam, idParam))))
+      }
+
+      def mkGetJUnitMetadataDef(clSym: Symbol,
+          mlSymOption: Option[Symbol]): DefDef = {
+        val methods = jUnitAnnotatedMethods(clSym)
+        val mlMethods = mlSymOption.map(jUnitAnnotatedMethods)
+
+        def liftAnnotations(methodSymbol: Symbol): List[Tree] = {
+          val annotations = methodSymbol.annotations
+          def clone(t: Tree): Tree = {
+            t match {
+              case lit: Literal => lit
+
+              case Select(t2, name) => Select(clone(t2), name)
+
+              case ValDef(mod, name, tpt, rhs) =>
+                ValDef(mod, name, tpt, clone(rhs))
+
+              case _ => t
+            }
+          }
+          def liftList(params: List[Tree]): List[Tree] =
+            params.map(clone)
+
+          // Find ad report unsupported JUnit annotations
+          annotations.foreach {
+            case ann if ann.atp.typeSymbol == testAnnotationSymbol && ann.original.isInstanceOf[Block] =>
+              reporter.error(ann.pos, "@Test(timeout = ...) is not " +
+                "supported in Scala.js JUnit Framework")
+
+            case ann if ann.atp.typeSymbol == fixMethodOrderAnnotationSymbol =>
+              reporter.error(ann.pos, "@FixMethodOrder(...) is not supported " +
+                "in Scala.js JUnit Framework")
+
+            case _ => // all is well
+          }
+
+          // Collect lifted representations of the JUnit annotations
+          annotations.collect {
+            case ann if annotationWhiteList.contains(ann.tpe.typeSymbol) =>
+              val args = if (ann.args != null) liftList(ann.args) else Nil
+              mkNewInstance(TypeTree(ann.tpe), args)
+          }
+        }
+
+        def defaultMethodMetadata(tpe: TypeTree)(mtdSym: MethodSymbol): Tree = {
+          val annotations = liftAnnotations(mtdSym)
+          mkNewInstance(tpe, List(
+              Literal(Constant(mtdSym.nameString)),
+              mkList(annotations)))
+        }
+
+        def mkList(elems: List[Tree]): Tree = {
+          val array = ArrayValue(TypeTree(definitions.ObjectTpe), elems)
+          val castArray = gen.mkCast(array, typeOf[Array[Object]])
```

Comment:
I reverse engineered it from some examples of trees that I had. But it also works without the cast. I will remove it.